### PR TITLE
[Leo] Add hexadecimal, octal, binary numerals.

### PR DESCRIPTION
**Pull request by acoglio**
This passes all the usual ACL2 checks.

### CODE REVIEW

File: leo.abnf [hunk 1]
```diff
@@ -150,17 +150,36 @@ lowercase-letter = %x61-7A ; a-z
...
+                  / %x41-46 ; A-F (uppercase)
+                  ; no a-f (lowercase)
+
 identifier = letter *( letter / decimal-digit / "_" )
              ; but not a keyword or a boolean literal or aleo1...
 
-numeral = 1*( decimal-digit *"_" )
+binary-numeral = 1*( binary-digit *"_" )
+
+octal-numeral = 1*( octal-digit *"_" )
```
Comment (bendyarm) [inline leo.abnf:170]:
```suggestion
"0o" 1*( octal-digit *"_" )
```

File: leo.abnf [hunk 2]
```diff
@@ -150,17 +150,36 @@ lowercase-letter = %x61-7A ; a-z
...
 
+hexadecimal-digit = decimal-digit
+                  / %x41-46 ; A-F (uppercase)
+                  ; no a-f (lowercase)
+
 identifier = letter *( letter / decimal-digit / "_" )
              ; but not a keyword or a boolean literal or aleo1...
 
-numeral = 1*( decimal-digit *"_" )
+binary-numeral = 1*( binary-digit *"_" )
```
Comment (bendyarm) [inline leo.abnf:168]:
```suggestion
binary-numeral = "0b" 1*( binary-digit *"_" )
```

File: leo.abnf [hunk 3]
```diff
@@ -150,17 +150,36 @@ lowercase-letter = %x61-7A ; a-z
...
+
+octal-numeral = 1*( octal-digit *"_" )
+
+decimal-numeral = 1*( decimal-digit *"_" )
+
+hexadecimal-numeral = 1*( hexadecimal-digit *"_" )
```
Comment (bendyarm) [inline leo.abnf:174]:
```suggestion
hexadecimal-numeral = "0x" 1*( hexadecimal-digit *"_" )
```

**Review by bendyarm — CHANGES_REQUESTED** [review]
Added some suggested changes.
Also we should think more about how to handle negative literals such as i8
other than field/group/scalar.
If we are handling then as unary minus operator, it looks to me like the grammar doesn't support that.

**Comment by acoglio** [comment]
Group coordinates are still in use.

File: leo.abnf
Comment (bendyarm) [inline leo.abnf:1]: Probably
```
cast-expression = unary-expression %s"as" named-primitive-type
```
should be changed to
```
cast-expression = unary-expression
                / unary-expression %s"as" named-primitive-type
```

Comment (acoglio) [reply leo.abnf:1]: Yes, there is a missing case in that rule. Great catch!

**Review by bendyarm — APPROVED** [review]
Looks good!

File: leo.abnf [hunk 4]
```diff
@@ -375,7 +394,8 @@ unary-expression = postfix-expression
                  / "!" unary-expression
                  / "-" unary-expression
 
-cast-expression = unary-expression %s"as" named-primitive-type
+cast-expression = unary-expression
```
[the remark this answers is not in the document]
Comment (acoglio) [reply leo.abnf:397]: Yes, an oversight indeed. It was not following the same hierarchical pattern as the other kinds of expressions.

**Review by d0cd — APPROVED** [review]
LGTM!

**Pull request merged**